### PR TITLE
Do no run Migration if Snapshot is the same as current version.

### LINKED ISSUE #15678: Do not always run migration
**Description**

Right now we run migrations every time we restart or change a transition, this can become quite a problem if we add more and more migrations, especially since it is often not easy to determine whether migration needs to be run or not (not based on the data we store in the column families).

The newest example is https://github.com/camunda/zeebe/pull/15443 where we always run the migration on failed jobs, to create missing job backoff entries.

I think this can be easily avoided if we simply write the Broker version into the state when this state was last time touched. 

Example:

Running at 8.3.5 we write into the state (default CF) the version 8.3.5. When upgrading we check the version in the state against our current broker version, if we see a mismatch we execute the migration if they are equal we don't need to run the migrations.

This can be done in the MigrationTransitionStep https://github.com/camunda/zeebe/blob/main/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/steps/MigrationTransitionStep.java

If we find no version, we know we need to migrate as well. After migration we always write the newest version to the state.

Additionally: We can avoid to start the follower with a newer snapshot, this might corrupt the data of the state and currently we have no way to detect this, but with the version we could stop the transition.

### PR DESCRIPTION
**Pull request by rodrigo-lourenco-lopes**
## Description

We don't run the Migration if the snapshot version is the same as the current version.

## Related issues

closes https://github.com/camunda/zeebe/issues/15678

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigratorImpl.java
Comment (oleschoenburg) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigratorImpl.java:79]: :upside_down_face: Feel free to just inline this variable here

**Review by lenaschoenburg — APPROVED** [review]
:tada: Looks good to me.

I'd like to get an additional review by ZPA just for awareness and to double check that we aren't missing anything. Maybe @korthout is interested again?

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/migration/DbMigratorImplTest.java
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/migration/DbMigratorImplTest.java:167]: ❌ I'm missing a test that shows what happens when the snapshot is newer than the current version

**Review by korthout — COMMENTED** [review]
Thanks @rodrigo-lourenco-lopes and @oleschoenburg 

👍 I like where this is going.

💭 It would be very useful to have some documentation about migrations. 

For example, when do they run? This used to be on all transitions: becoming follower, on becoming leader, etc, but now only one time on the first transition.

And on what data do they run? I.e. it differs per snapshot version. What does this mean for the way event appliers must be written. For example, for multi-tenancy we moved data in a migration task to other column families without introducing new record versions and new event appliers. Instead, we adjusted what the event appliers do. Normally, event appliers must deterministically produce the same state changes every time they are applied, but that was [broken for multi-tenancy migrations](https://camunda.slack.com/archives/C0555930H7F/p1694682904830279?thread_ts=1694081342.937559&cid=C0555930H7F).

At this time, I find it hard to reason about the case where a newer snapshot is received and the broker is transitioning from inactive to follower. Does the migration fail because Incompatible or does it try to run the migrations even if they are not complete?

It would be good to document (perhaps in tests) what all combinations are between:
- differences between snapshot and broker version
- when the migration is run (on which role transition)

❌ At the very least, I see a need for the case where the snapshot is newer than the broker

**Comment by oleschoenburg** [comment]
> At this time, I find it hard to reason about the case where a newer snapshot is received and the broker is transitioning from inactive to follower. Does the migration fail because Incompatible or does it try to run the migrations even if they are not complete?

I'm not sure if I fully understand your question but the behavior on any transition (other than transitioning to `INACTIVE`) is to check if the snapshot version is compatible with the current broker version (i.e. same version or a supported upgrade path from a previous patch or minor). The new behavior here is to just not run migrations if the version did not change, i.e. all of the migrations already successfully once.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/migration/DbMigratorImplTest.java
Comment (oleschoenburg) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/migration/DbMigratorImplTest.java:167]: This test should have been added in https://github.com/camunda/zeebe/pull/16421 then, the behavior did not change with this PR.

Comment (korthout) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/migration/DbMigratorImplTest.java:167]: No tests were added that show that these lines throw exceptions

https://github.com/camunda/zeebe/blob/d1c3915ae4bba2f60d10ca611f1670449f519633/zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigratorImpl.java#L104-L109

Comment (oleschoenburg) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/migration/DbMigratorImplTest.java:167]: I'll add a follow to https://github.com/camunda/zeebe/issues/15885 for this then, I think this shouldn't block this PR because it's not new behavior.

**Pull request merged**
